### DISCUSSION TITLE
New driver

### DISCUSSION THREAD

**Discussion by HondaRulez** [olikraus/u8g2#1767] [General]
Hi,
any docs about how to add a new driver ?

I have a 170x40 VFD graphic display, the display have parellel interface similar to a SRAM or so.. address, data also have some other i/o.
currently due the limited I/O pins on the host side I made a I2C interface.
the display works perfectly.
My goal is using U8g2 to draw items and use different sizes of the fonts on this display.
so I would add this one to the drivers and use the lib as usual on other devices.

Thanks.

**Comment by olikraus** [comment]
What is your exact type of VFD?
U8g2 beta release contains support for a GU800 128x64 VFD, see also here: https://github.com/olikraus/u8g2/issues/1604

``` 
U8G2_GU800_128X64_1_4W_SW_SPI u8g2(U8G2_R0, /* clock=*/ 13, /* data=*/ 11, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
U8G2_GU800_128X64_1_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
```

You can download the latest U8g2 beta release from here: https://github.com/olikraus/U8g2_Arduino/archive/master.zip
Arduino IDE:
1. Remove the existing U8g2_Arduino library (https://stackoverflow.com/questions/16752806/how-do-i-remove-a-library-from-the-arduino-environment)
2. Install the U8g2_Arduino Zip file via Arduino IDE, add zip library menu (https://www.arduino.cc/en/Guide/Libraries).

PlatformIO:
`platformio.ini` (https://docs.platformio.org/en/latest/projectconf/section_env_library.html#lib-deps) should include 
```
lib_deps =
  u8g2=https://github.com/olikraus/U8g2_Arduino/archive/master.zip
```

**Reply by HondaRulez** [reply]
The driver will not work due my own intarface (i2c to parallel conversion)
BTW I using platformio and ESP32 / IDF with this library.

[VFD.pdf](https://github.com/olikraus/u8g2/files/7966201/VFD.pdf)

The VFD's driver chip is iItron sla904SF

**Reply by olikraus** [reply]
hmm... difficult. how does your i2c to parallel conversion look like?

**Reply by HondaRulez** [reply]
using 2 x MCP23017 I/O expander for address, data, and page select lines, RD & WR linesr directly driven from the ESP32,
Really some directions where to start would be enought to me, Previously I used alot of LCDs with own drivers, so the programming and electronic is not a problem to me at all..,I need to know how the library is work..  at the end I just need to map a memory buffer to the LCD memory ?, the controller is really dumb, just mapping a bitbuffer to the display, nothing else.. what if I make a driver file similar to the existing one, where can I register my functions to be used by the lib ?  and so on...

**Reply by olikraus** [reply]
I think you could use the A2Printer as a template.
So setup might look like this:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8g2_d_setup.c#L6606-L6613

In your case it will look more like this:

```
u8g2_SetupDisplay(u8g2, u8x8_d_vfd_170x40, u8x8_cad_empty, u8x8_cad_empty, u8x8_cad_empty);
```

the procedure `u8x8_d_vfd_170x40` has to be provided by you (see below).

The internal memory allocation could be:
```
buf = u8g2_m_22_9_f(&tile_buf_height);
```

For the device `u8x8_d_vfd_170x40` you could takeover the code from here:

https://github.com/olikraus/u8g2/blob/master/csrc/u8x8_d_a2printer.c

The function prototype will be:

uint8_t u8x8_d_vfd_170x40u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)

Compare here:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8_d_a2printer.c#L165

This function will receive several messages via the "msg" argument. You can ignore most messages, but you need to handle `U8X8_MSG_DISPLAY_SETUP_MEMORY` like in the following code:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8_d_a2printer.c#L169-L171

The only purpose of this SETUP_MEMORY request is to provide the display info structure:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8_d_a2printer.c#L139

Again most elements can be ignored, but you need to assign tile and pixel dimensions so that u8g2 knows the size of your display.

The other important message is `DRAW_TILE`:

https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8_d_a2printer.c#L84

`DRAW_TILE` is described here:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8.h#L491-L514

Finally Arduino setup will look like this:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/sys/arduino/u8g2_page_buffer/A2Printer/A2Printer.ino#L106-L115

ok, that was a quick introduction. I am sure you will have many questions...

**Reply by HondaRulez** [reply]
I just took a look, it's not so difficult :) My 1st question is about the user side initialisation, because I using both I2C and GPIO, at least 1 GPIO need to be defined by the user, I seen some driver uses GPIO together with SPI bus named "CD" but did not seen together with I2C bus, is it posibble to getting a GPIO definition from the user ? or need to be hardcocded in my case ?

**Reply by olikraus** [reply]
U8g2 provides a communication interface for you, however your case I assume it is better to hardcode everything: This means you can ignore all kind of GPIO assignments.
For example in  the A2 driver you can see this command to transfer the bitmap data to the display:
https://github.com/olikraus/u8g2/blob/ef9fa9f8f8aaa7767c7dd959dcf512433ddc2047/csrc/u8x8_d_a2printer.c#L109

In your case you need to write a replacement for `u8x8_cad_SendData` which transfers data via your i2c interface ot the VFD.
As decribed above the x y position is stored in ((u8x8_tile_t *)arg_ptr) if you receive the DRAW_TILE message. From the x y pos you need to calculate the address of the data byte location in the VFD.

**Reply by HondaRulez** [reply]
ok, some files are "generated" like u8g2_d_memory.c and u8g2_d_setup.c, so I need to edit a template somwhere or I can edit directly these files ? or how the file generation works ?

**Reply by olikraus** [reply]
> ok, some files are "generated" like u8g2_d_memory.c and u8g2_d_setup.c, so I need to edit a template somwhere or I can edit directly these files ? or how the file generation works ?

Code Generation is not required if you follow my suggestions from above.
My suggested procedure `u8g2_m_22_9_f` already exists but is bigger than required. The code generatore would probably generate a new procedure `u8g2_m_22_5_f` which would fit better, but you could also write this `u8g2_m_22_5_f` by your own.

Also the setup procedure can be written by yourself. So, no need to generate anything.

**Reply by HondaRulez** [reply]
sorry for the dumb questions :) I just like to knows the backgrounds what I doing :)
naming scheme...

uint8_t *u8g2_m_22_9_1(uint8_t *page_cnt)
{
  #ifdef U8G2_USE_DYNAMIC_ALLOC
  *page_cnt = 1;
  return 0;
  #else
  static uint8_t buf[176];
  *page_cnt = 1;
  return buf;
  #endif
}
this functions allocate only 176 bytes (I expected 198)

also what means the 22 the 9 and the 'f'

checking some other drivers I would declare 170_5_4

because the LCD have 4 individual display pages and 1 data byte covers 1 column and 8 lines.

**Reply by olikraus** [reply]
The code generate creates those functions, based on the size of the displays, so each display will have its own memory allocation function. Of course those functions are reused as much as possible (so two displays with the same dimension will share the same memory function).

The memory function look like this: u8g2_m_22_9_f

22: The width of the display in tiles ( (width_in_pixel + 7) / 8) (176 pixel in your case)
9: Tile height ( (height_in_pixel + 7) / 8 )

The f at the end tells u8g2 the type of data upload to the display:
1: only one page can be rendered and uploaded to the display (picture loop, slower but lesser RAM required)
f: the full display memory is mirrored in u8g2. User can use the send command to upload the complete data to the display (faster, but more RAM required)

A page is defined as one tile row, which is 8 pixel height and 170 pixel width in your case 
The number of pages has to be returned in *page_cnt
So in your case, if you want all your graphics beeing rendered in the uC you need to set *page_cnt = 5;
(5*8=40 pixel) 

Lets make more practical examples:
*page_cnt = 1 (one page of display memory in uC) --> 176 bytes required
*page_cnt = 5 (full display memory in uC) --> 880 bytes requred

There are some programming differences if your use this full memory approach vs. one page approach (see U8g2 Manual).
